GCP Cloud Spending Analysis

1. Hanina Marchia Putri (5027261012)                                                                                                                   
2. Marchelio Fahrul Rahmansyah (5027261027)                                                                                                            
3. Muhammad Affan Arsyad (5027261081)

Topik Yang dipih : Cloud Computing

Komputasi awan seperti Google Cloud Platform (GCP) memungkinkan pengguna menyewa sumber daya komputasi, penyimpanan, dan jaringan dengan model bayar sesuai pemakaian. Model ini fleksibel, tetapi biayanya sulit diprediksi karena bergantung pada jenis layanan, region, volume penggunaan, dan durasi pemakaian. Sumber daya juga sering tidak dimanfaatkan optimal, misalnya utilisasi CPU dan memori rendah tetapi tetap ditagih penuh. Karena itu, pengelolaan biaya awan perlu didukung analisis berbasis data.
Dataset dalam proyek ini memuat 1.000 catatan penggunaan GCP pada 30 Juni sampai 29 Agustus 2024, mencakup 25 layanan di 12 region. Variabelnya meliputi kuantitas penggunaan, utilisasi CPU dan memori, trafik jaringan, durasi, dan biaya. Melalui eksplorasi data (EDA), proyek ini bertujuan memahami distribusi variabel, menemukan pola dan pencilan, serta melihat faktor yang memengaruhi biaya layanan awan.

Question:
1. Region mana yang paling mahal, dan apakah selisihnya jauh dengan negara negara yang lain?
2. Berapa rata-rata dan standar deviasi dari CPU Utilization? Berapa koefisien variasinya?
3. Layanan (Service Name) mana yang rata-rata biayanya paling tinggi dan paling bervariasi (std/CV)?

In [1]:
import pandas as pd

df = pd.read_csv(r"C:\Users\Hanina Marchia\Project 1 EDA Statprob\gcp_final_approved_dataset.csv")
print(df.shape)
df.head()

(1000, 15)


,Resource ID,Service Name,Usage Quantity,Usage Unit,Region/Zone,CPU Utilization (%),Memory Utilization (%),Network Inbound Data (Bytes),Network Outbound Data (Bytes),Usage Start Date,Usage End Date,Cost per Quantity ($),Unrounded Cost ($),Rounded Cost ($),Total Cost (INR)
0,res-ST6BAJ2N,Cloud Dataproc,954.9843,Requests,europe-north1,92.72,70.42,77097035547,8.268593e+10,01-08-2024 22:24,07-08-2024 06:54,5.24,5004.12,5004,415332
1,res-TYGNR0RV,Pub/Sub,479.6348,GB,europe-west1,37.78,46.80,8239357017,1.135149e+10,23-08-2024 09:18,25-08-2024 07:26,9.82,4710.01,4710,390930
2,res-S3I9C869,BigQuery,114.4129,GB,southamerica-east1,82.54,59.47,23358691883,2.345081e+10,18-07-2024 11:13,22-07-2024 00:45,7.37,843.22,843,69969
3,res-1RY9BZ6G,Cloud Endpoints,413.1930,GB,us-central1,86.68,97.63,53404385778,6.078019e+10,20-07-2024 00:15,21-07-2024 10:16,4.41,1822.18,1822,151226
4,res-S3HQYIZ2,Cloud Spanner,335.9892,GB,asia-southeast1,82.95,21.58,10795270461,1.416138e+10,21-08-2024 21:25,25-08-2024 00:21,6.09,2046.17,2046,169818


1. Sumber data:https://www.kaggle.com/datasets/sairamn19/gcp-cloud-billing-data/data. Lisensi: Community Data License Agreement - Sharing - Version 1.0 (https://cdla.io/sharing-1-0/)                                                                                                              
2. Pengumpul Data: Sairam N, Waktu Pengumpulan: Diunggah ke platform Kaggle pada tahun 2020., Cara Pengumpulan: Data dikumpulkan melalui fitur Cloud Billing Export bawaan dari Google Cloud Platform (GCP). Fitur ini mengekspor catatan transaksi dan detail konsumsi resource secara otomatis dari GCP Billing ke BigQuery/CSV, yang mencakup data pemakaian layanan, waktu, lokasi pusat data, serta rincian biaya.                                    
3. Jumlah Baris: 1.000 baris record data, Jumlah Kolom: 15 kolom.
4. Kamus Data
| Kolom | Arti | Jenis Variabel | Skala Pengukuran | Satuan | Contoh |
|---|---|---|---|---|---|
| Resource ID | ID unik untuk mengidentifikasi resource GCP | Kualitatif | Nominal | Tidak ada | res-ST6BAJ2N |
| Service Name | Nama layanan Google Cloud yang digunakan | Kualitatif | Nominal | Tidak ada | Cloud Spanner |
| Usage Quantity | Jumlah penggunaan suatu layanan atau resource | Kuantitatif | Rasio | Sesuai Usage Unit | 335.9892 |
| Usage Unit | Satuan yang digunakan untuk mengukur penggunaan resource | Kualitatif | Nominal | Tidak ada | GB |
| Region/Zone | Region atau zona tempat resource digunakan | Kualitatif | Nominal | Tidak ada | asia-southeast1 |
| CPU Utilization | Persentase penggunaan CPU pada resource | Kuantitatif | Rasio | % | 82.95% |
| Memory Utilization | Persentase penggunaan memori pada resource | Kuantitatif | Rasio | % | 21.58% |
| Network Inbound Data | Jumlah data yang masuk ke resource melalui jaringan | Kuantitatif | Rasio | Byte | 77097035547 |
| Network Outbound Data | Jumlah data yang keluar dari resource melalui jaringan | Kuantitatif | Rasio | Byte | 82685930000 |
| Usage Start Date | Tanggal dan waktu mulai penggunaan resource | Kuantitatif | Interval | Tanggal & waktu | 01-08-2024 22:24 |
| Usage End Date | Tanggal dan waktu berakhirnya penggunaan resource | Kuantitatif | Interval | Tanggal & waktu | 07-08-2024 06:54 |
| Cost per Quantity | Biaya untuk setiap satuan penggunaan resource | Kuantitatif | Rasio | USD | 5.24 |
| Unrounded Cost | Total biaya penggunaan sebelum pembulatan | Kuantitatif | Rasio | USD | 5004.12 |
| Rounded Cost | Total biaya penggunaan setelah pembulatan | Kuantitatif | Rasio | USD | 5004 |
| Total Cost | Total biaya penggunaan resource dalam mata uang rupee India | Kuantitatif | Rasio | INR | 415332 |

In [2]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 15 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Resource ID                    1000 non-null   str    
 1   Service Name                   1000 non-null   str    
 2   Usage Quantity                 1000 non-null   float64
 3   Usage Unit                     1000 non-null   str    
 4   Region/Zone                    1000 non-null   str    
 5   CPU Utilization (%)            1000 non-null   float64
 6   Memory Utilization (%)         1000 non-null   float64
 7   Network Inbound Data (Bytes)   1000 non-null   int64  
 8   Network Outbound Data (Bytes)  1000 non-null   float64
 9   Usage Start Date               1000 non-null   str    
 10  Usage End Date                 1000 non-null   str    
 11  Cost per Quantity ($)          1000 non-null   float64
 12  Unrounded Cost ($)             1000 non-null   float64
 13  

In [3]:
df.isna().sum()           # jumlah data kosong per kolom

Resource ID                      0
Service Name                     0
Usage Quantity                   0
Usage Unit                       0
Region/Zone                      0
CPU Utilization (%)              0
Memory Utilization (%)           0
Network Inbound Data (Bytes)     0
Network Outbound Data (Bytes)    0
Usage Start Date                 0
Usage End Date                   0
Cost per Quantity ($)            0
Unrounded Cost ($)               0
Rounded Cost ($)                 0
Total Cost (INR)                 0
dtype: int64

In [4]:
df.duplicated().sum()     # jumlah baris duplikat

np.int64(0)

In [5]:
print('Variabel Numerik')
hasil = pd.DataFrame({
    "Usage Quantity": [
        df["Usage Quantity"].mean(),
        df["Usage Quantity"].median(),
        df["Usage Quantity"].mode()[0],
        df["Usage Quantity"].var(),
        df["Usage Quantity"].std(),
        df["Usage Quantity"].quantile(0.25),
        df["Usage Quantity"].quantile(0.75)
    ],
    
    "CPU Utilization (%)": [
        df["CPU Utilization (%)"].mean(),
        df["CPU Utilization (%)"].median(),
        df["CPU Utilization (%)"].mode()[0],
        df["CPU Utilization (%)"].var(),
        df["CPU Utilization (%)"].std(),
        df["CPU Utilization (%)"].quantile(0.25),
        df["CPU Utilization (%)"].quantile(0.75)
    ],
    
    "Total Cost (INR)": [
        df["Total Cost (INR)"].mean(),
        df["Total Cost (INR)"].median(),
        df["Total Cost (INR)"].mode()[0],
        df["Total Cost (INR)"].var(),
        df["Total Cost (INR)"].std(),
        df["Total Cost (INR)"].quantile(0.25),
        df["Total Cost (INR)"].quantile(0.75)
    ]
}, index=[ #nampilin nama baris
    "Mean",
    "Median",
    "Mode",
    "Variance",
    "Standard Deviation",
    "Q1",
    "Q3"
])

hasil

Variabel Numerik


,Usage Quantity,CPU Utilization (%),Total Cost (INR)
Mean,383.343530,50.943680,1.772228e+05
Median,336.721800,52.030000,1.038330e+05
Mode,91.866700,0.690000,2.631100e+04
Variance,89889.431411,830.865741,2.982042e+10
Standard Deviation,299.815662,28.824742,1.726859e+05
Q1,106.818150,26.740000,4.154150e+04
Q3,616.048825,75.647500,2.703725e+05


In [6]:
df["Region/Zone"].value_counts()

Region/Zone
asia-east1                 103
europe-west1                99
asia-southeast1             91
southamerica-east1          89
europe-north1               83
northamerica-northeast1     82
us-east1                    82
us-central1                 81
europe-west3                77
asia-northeast1             74
australia-southeast1        71
us-west1                    68
Name: count, dtype: int64

In [7]:
df["Region/Zone"].value_counts(normalize=True) * 100

Region/Zone
asia-east1                 10.3
europe-west1                9.9
asia-southeast1             9.1
southamerica-east1          8.9
europe-north1               8.3
northamerica-northeast1     8.2
us-east1                    8.2
us-central1                 8.1
europe-west3                7.7
asia-northeast1             7.4
australia-southeast1        7.1
us-west1                    6.8
Name: proportion, dtype: float64

In [ ]:
import matplotlib.pyplot as plt

plt.hist(df["CPU Utilization (%)"], bins=20)
plt.title("Sebaran CPU Utilization")
plt.xlabel("CPU Utilization (%)")
plt.ylabel("Frekuensi")
plt.show()

Frekuensi di hampir semua interval (dari 0% sampai 100%) berada pada kisaran yang mirip, yaitu sekitar 40–60.
Tidak ada penumpukan data yang jelas di salah satu sisi (kiri atau kanan) dengan ekor panjang ke arah lainnya.
Puncak-puncak kecil justru muncul di beberapa titik (sekitar 5-10%, 55-60%, dan 95-100%), namun sebarannya tetap relatif merata di sepanjang rentang 0-100%.

In [ ]:
import matplotlib.pyplot as plt

plt.hist(df["Memory Utilization (%)"], bins=20)
plt.title("Sebaran Memory Utilization")
plt.xlabel("Memory Utilization (%)")
plt.ylabel("Frekuensi")
plt.show()

Frekuensi tersebar cukup merata di seluruh rentang 0–100%, berkisar antara 40-60.
Tidak terlihat penumpukan data yang signifikan di salah satu ujung dengan ekor panjang ke arah sebaliknya, sehingga tidak bisa dikatakan miring ke kanan atau miring ke kiri.
Ada sedikit variasi naik turun (misalnya lonjakan di sekitar 30-35%, 50-55%, dan 90-100%), tapi ini lebih menyerupai fluktuasi acak dalam data daripada pola kemiringan yang konsisten.

In [ ]:
plt.boxplot(df["Rounded Cost ($)"])
plt.title("Boxplot Rounded Cost")
plt.ylabel("Cost (USD)")
plt.show()

Boxplot menunjukkan adanya pencilan pada Rounded Cost ($), terutama pada nilai yang tinggi. Pencilan ditunjukkan oleh titik-titik yang berada di atas whisker dan melebihi batas atas Q3 + 1,5 × IQR. Hal ini menunjukkan adanya beberapa resource dengan biaya yang jauh lebih besar dibandingkan sebagian besar resource lainnya.

In [ ]:
df["Region/Zone"].value_counts().plot(kind="bar")
plt.title("Jumlah Data per Region/Zone")
plt.xlabel("Region/Zone")
plt.ylabel("Jumlah")
plt.show() 

Jumlah data pada setiap Region/Zone cukup bervariasi. asia-east1 memiliki jumlah data terbanyak, sedangkan us-west1 memiliki jumlah data paling sedikit. Meskipun terdapat perbedaan jumlah data antar-Region/Zone, tidak ada satu Region/Zone yang mendominasi sebagian besar data.

KESIMPULAN

Jawaban Soal Sel 2:
1. Region mana yang paling mahal, dan apakah selisihnya jauh dengan negara negara yang lain?
Region europe-north1 dengan biaya 5004$, selisihnya tidak terlalu signifikan jika dibandingankan dengan europe-west1, tetapi akan terlihat signifikan jika dibandingkan dengan yang lainnya (selain europe-west1)
2. Layanan (Service Name) mana yang rata-rata biayanya paling tinggi dan paling bervariasi (std/CV)?
Layanan dengan rata-rata biaya paling tinggi adalah Cloud NAT 3388.69, dan untuk yang paling bervariasi Cloud Dataproc 93,58%
3. Berapa rata-rata dan standar deviasi dari CPU Utilization? Berapa koefisien variasinya?
Rata-rata CPU Utilization: 50.94%,
Standar deviasi CPU Utilization: 28.82%,
Koefesien variasinya: 56.58%

Temuan paling menarik:
1. Jumlah catatan per region tidak selalu sejalan dengan rata-rata biayanya. asia-east1 memiliki catatan terbanyak (103), tetapi rata-rata biaya per catatannya paling rendah (144.325,72 INR). Rata-rata tertinggi ada di us-east1, sebesar 218.248,50 INR dari 82 catatan.
2. Utilisasi rendah bisa muncul bersama biaya tinggi. Dari 26 catatan dengan CPU dan memori di bawah 20%, ada 8 catatan dengan biaya minimal 270.372,50 INR. Salah satu contohnya adalah Cloud Data Fusion dengan utilisasi CPU 7,96%, utilisasi memori 10,82%, dan biaya 721.519 INR. Data ini belum menjelaskan penyebab biayanya.
3. Modus dari CPU Utilization (%) dan Usage Quantity bersifat multimodal, yang artinya variabel tersebut memiliki beberapa data dengan frekuensi yang sama. CPU Utilization memiliki modus hingga 47 dengan frekuensi 2 kali, sedangkan Usage Quantity memiliki 2 modus dengan frekuensi 2 kali. Pada tabel variabel numerik, modus yang ditampilkan adalah modus pertama atau yang paling kecil.

Keterbatasan data:
1. Cakupan Waktu yang Singkat
2. Informasi Konversi Mata Uang Tidak Rinci
3. Ketiadaan Metadata & Dokumentasi Sumber

Ide pertanyaan lanjutan yang ingin dianalisis:
1. Apakah penggunaan CPU dapat memprediksi biaya resource? (Uji Regresi)
2. Apakah rata-rata biaya penggunaan berbeda antar layanan GCP? (Uji Hipotesis)
3. Apakah penggunaan CPU berkaitan dengan biaya resource? (Uji Korelasi)

In [ ]:
ringkasan_biaya = df.groupby("Service Name")["Rounded Cost ($)"].agg(
    rata2="mean",
    std="std",
    jumlah_data="count" 
)

ringkasan_biaya["CV (%)"] = (ringkasan_biaya["std"] / ringkasan_biaya["rata2"]) * 100
ringkasan_biaya = ringkasan_biaya.round(2)
ringkasan_biaya

In [ ]:
rata2_cpu = df["CPU Utilization (%)"].mean()
std_cpu = df["CPU Utilization (%)"].std()
cv_cpu = (std_cpu / rata2_cpu) * 100


print(f"Rata-rata CPU Utilization: {rata2_cpu:.2f}%")
print(f"Standar deviasi CPU Utilization: {std_cpu:.2f}%")
print(f"Koefesien variasinya: {cv_cpu:.2f}%")

## Referensi dan Catatan Penggunaan AI

| Alat | Dipakai untuk | Yang kami pelajari |
|---|---|---|
| Claude AI | Menanyakan syntax-syntax yang tidak kami ketahui | Mempelajari syntax yang masih belum pernah digunakan |
| Claude AI | Menanyakan bagaimana cara codingannya bekerja | Mempelajari bagaimana codingan bekerja |
| Claude AI | Menanyakan kenapa codingan yang dibuat error | Mengetahui letak kesalahan codingan |
| ChatGPT | Menanyakan arti error 'unicodeescape codec can't decode bytes' saat memanggil data | Memakai r (raw string) sebelum tanda kutip agar Python membaca backslash di lokasi file sebagai karakter biasa, bukan escape character. Bisa juga dengan mengganti \ menjadi / |
| ChatGPT | Cara membuat tabel di markdown | Menggunakan format \|...\| dan \|---\| untuk mengatur alignment |
| ChatGPT | Cara membuat tabel di code | Menggunakan pd.DataFrame(data) dan menyimpan tabel di hasil untuk bisa ditampilkan |